File: backend/app/models/challenge.py
from sqlalchemy import Column, Integer, String, Float, DateTime, JSON, ForeignKey, Boolean
from sqlalchemy.sql import func
from app.db.base import Base

class Challenge(Base):
    __tablename__ = "challenges"
    
    id = Column(Integer, primary_key=True, index=True)
    challenger_id = Column(Integer, ForeignKey("users.id"))
    challengee_id = Column(Integer, ForeignKey("users.id"))
    
    type = Column(String)  # singles, doubles, team
    game_type = Column(String)  # 8ball, 9ball, straight
    race_type = Column(String)  # standard, handicap
    race_to = Column(JSON)  # {"challenger": 5, "challengee": 5}
    
    status = Column(String)  # pending, accepted, declined, completed
    result = Column(JSON, nullable=True)
    
    # Money match details
    is_money_match = Column(Boolean, default=False)
    stake_amount = Column(Float, default=0)
    side_bets = Column(JSON, default=[])
    
    scheduled_for = Column(DateTime(timezone=True), nullable=True)
    completed_at = Column(DateTime(timezone=True), nullable=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())

File: app/models/challenge.js
export class Challenge {
    constructor(data = {}) {
        this.id = data.id || '';
        this.challengerId = data.challengerId || '';
        this.challengerName = data.challengerName || '';
        this.challengerPhoto = data.challengerPhoto || '';
        this.challengeeId = data.challengeeId || '';
        this.challengeeName = data.challengeeName || '';
        this.challengeePhoto = data.challengeePhoto || '';
        this.status = data.status || 'pending'; // pending, accepted, declined, completed
        this.type = data.type || 'singles'; // singles, doubles, team
        this.gameType = data.gameType || '8ball'; // 8ball, 9ball, straight
        this.raceType = data.raceType || 'standard'; // standard, handicap
        this.race = {
            challenger: data.race?.challenger || 5,
            challengee: data.race?.challengee || 5
        };
        
        // Money management
        this.moneyMatch = {
            enabled: data.moneyMatch?.enabled || false,
            amount: data.moneyMatch?.amount || 0,
            sideBets: data.moneyMatch?.sideBets || [], // Additional bets during match
            settled: data.moneyMatch?.settled || false,
            settledTimestamp: data.moneyMatch?.settledTimestamp || null,
            settledBy: data.moneyMatch?.settledBy || null // Player who confirmed settlement
        };

        this.createdAt = data.createdAt || new Date();
        this.scheduledFor = data.scheduledFor || null;
        this.result = data.result || null;
        this.notes = data.notes || '';
    }

    addSideBet(bet) {
        if (!this.moneyMatch.enabled) return false;
        
        this.moneyMatch.sideBets.push({
            ...bet,
            timestamp: new Date(),
            settled: false
        });
        return true;
    }

    calculateTotalStake() {
        if (!this.moneyMatch.enabled) return 0;
        
        const mainStake = this.moneyMatch.amount;
        const sideBetsTotal = this.moneyMatch.sideBets.reduce((total, bet) => total + bet.amount, 0);
        
        return mainStake + sideBetsTotal;
    }

    settleMatch(settlerId) {
        if (!this.moneyMatch.enabled || this.moneyMatch.settled) return false;
        
        this.moneyMatch.settled = true;
        this.moneyMatch.settledTimestamp = new Date();
        this.moneyMatch.settledBy = settlerId;
        return true;
    }

    getMoneyMatchSummary() {
        if (!this.moneyMatch.enabled) return null;

        return {
            mainStake: this.moneyMatch.amount,
            sideBets: this.moneyMatch.sideBets,
            totalStake: this.calculateTotalStake(),
            settled: this.moneyMatch.settled,
            settledTimestamp: this.moneyMatch.settledTimestamp,
            settledBy: this.moneyMatch.settledBy
        };
    }
}